### PR TITLE
[SI-4691, SI-6008] improve patmat analyses: irrefutable user-defined extractors, no-op type tests

### PR DESCRIPTION
**Pull request by adriaanm**
review by @retronym

SI-4691 exhaustivity: `unapply: Some` = irrefutable
   overhauls treeMakersToConds in the process -- was getting a bit unwieldy

SI-6008 use static knowledge of success of type tests
augment the equality axioms to take into account that
a type test against the static type of a variable succeeds unless the variable is null

for exhaustivity we disregard null, so the type test always succeeds
during unreachability we model this knowledge as the obvious implication

ps: -Ypatmat-debug is a no-op for now, since it looks like the inliner fails to inline patmatDebug (/cc @magarciaEPFL)

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -184,7 +186,7 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
       // (that would require more sophistication when generating trees,
       //  and the only place that emits Matches after typers is for exception handling anyway)
       if(phase.id >= currentRun.uncurryPhase.id) debugwarn("running translateMatch at "+ phase +" on "+ selector +" match "+ cases)
-      // patmatDebug("translating "+ cases.mkString("{", "\n", "}"))
+      // patmatDebug ("translating "+ cases.mkString("{", "\n", "}"))
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:190]: What's with all these single-space additions? That's not cool to bundle with substantive changes.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:190]: it's actually in a separate commit, that only does that

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:190]: oh, I didn't spot the space between the patmatDebug and the opening paren (search/replace mistake..)
sorry about that, but you can completely ignore this commit except for the TODO comment at the top

**Comment by adriaanm** [comment]
I rewrote the commits to drop the annoying whitespace in the origin -Ypatmat-debug commit

the following diffs are all empty:
git diff daaa4488bd..54b26064d9 -w
git diff 9c09d32971..8234ba3905
git diff c3cb3e3a51..b61b5fffb6

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala (hunk 1 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:190]: Your contrition is deemed sufficient.

**Comment by retronym** [comment]
lgtm

**Pull request merged**